numba/np/arraymath.py

```diff
@@ -355,117 +354,133 @@ def array_sum_impl_axis(arr, axis):
 def array_cumsum(arr):
-    is_integer = arr.dtype in types.signed_domain | frozenset([types.bool_])
+    if isinstance(arr, types.Array):
+        is_integer = arr.dtype in types.signed_domain | frozenset([types.bool_])
 
-    def array_cumsum_impl(arr):
-        if is_integer:
-            dtype = np.intp
-        else:
-            dtype = arr.dtype
-        out = np.empty(arr.size, dtype)
-        c = np.zeros(1, dtype)[0]
-        for idx, v in enumerate(arr.flat):
-            c += v
-            out[idx] = c
-        return out
+        def array_cumsum_impl(arr):
+            if is_integer:
+                dtype = np.intp
+            else:
+                dtype = arr.dtype
+            out = np.empty(arr.size, dtype)
+            c = np.zeros(1, dtype)[0]
+            for idx, v in enumerate(arr.flat):
+                c += v
+                out[idx] = c
+            return out
 
-    return array_cumsum_impl
+        return array_cumsum_impl
 
 
 @overload(np.cumprod)
 @overload_method(types.Array, "cumprod")
 def array_cumprod(arr):
-    is_integer = arr.dtype in types.signed_domain | frozenset([types.bool_])
+    if isinstance(arr, types.Array):
+        is_integer = arr.dtype in types.signed_domain | frozenset([types.bool_])
 
-    def array_cumprod_impl(arr):
-        if is_integer:
-            dtype = np.intp
-        else:
-            dtype = arr.dtype
-        out = np.empty(arr.size, dtype)
-        c = np.ones(1, dtype)[0]
-        for idx, v in enumerate(arr.flat):
-            c *= v
-            out[idx] = c
-        return out
+        def array_cumprod_impl(arr):
+            if is_integer:
+                dtype = np.intp
+            else:
+                dtype = arr.dtype
+            out = np.empty(arr.size, dtype)
+            c = np.ones(1, dtype)[0]
+            for idx, v in enumerate(arr.flat):
+                c *= v
+                out[idx] = c
+            return out
 
-    return array_cumprod_impl
+        return array_cumprod_impl
 
 
 @overload(np.mean)
 @overload_method(types.Array, "mean")
 def array_mean(arr):
-    is_number = arr.dtype in types.integer_domain | frozenset([types.bool_])
-
-    def array_mean_impl(arr):
-        # Can't use the naive `arr.sum() / arr.size`, as it would return
-        # a wrong result on integer sum overflow.
-        if is_number:
-            dtype = types.float64
-        else:
-            dtype = arr.dtype
-        c = np.zeros(1, dtype)[0]
-        for v in np.nditer(arr):
-            c += v.item()
-        return c / arr.size
+    if isinstance(arr, types.Array):
+        is_number = arr.dtype in types.integer_domain | frozenset([types.bool_])
+
+        def array_mean_impl(arr):
+            # Can't use the naive `arr.sum() / arr.size`, as it would return
+            # a wrong result on integer sum overflow.
+            if is_number:
+                dtype = types.float64
+            else:
+                dtype = arr.dtype
```

Comment:
Move to outside of the impl. function

Reply:
Resolved in [4e38728](https://github.com/numba/numba/pull/8415/commits/4e38728908f9f1583ebd37e4b694f68d9f4c5c97)